### PR TITLE
Various PCRE syntax improvements

### PR DESCRIPTION
**Pull request by sfstewman**
Most of this was driven by a desire to make `cvtpcre` correctly handle parsing the PCRE test suite, and for it to correctly identify unsupported PCRE features so the output wouldn't have to be heavily hand-edited to eliminate things like backreferences.

### PCRE syntax:
- Handles `\x` codes without trailing hex digits
- Handles `\cX` control characters
- Handles `\b` and `\e` in character classes
- Handles PCRE comments `(?# ... )`
- Handles `\A` and `\z`
- Handles `\E` without a preceding `\Q`
- Handles partial support for named capture groups, although does not capture the name
- Repeats now follow PCRE rules, which are somewhat more strict

### Marks a lot of PCRE syntax as "unsupported" to give better error messages:
- Backreferences denote through `\1` .. `\7`, `\g`, and `\k`
- Lookahead/lookbehind assertions
- Non-greedy repeats (eg: `*?`) and atomic repeats (eg: `*+`) are marked as unsupported
- `\b`, `\B`, and `\G` are unsupported
- Various `(*VERB)`, `(*VERB:NAME)`, and `(*:NAME)` backtracking controls are unsupported
- `\K` "capture group reset" is (currently) unsupported
- `\Z` is (currently) unsupported

### Other related improvements:
- `fsm_example` outputs `\xHH` for unprintable characters
- Adds `M` records to `retest` to turn on/off regexp modifiers
- Fixes some bugs in `cvtpcre`
- `cvtpcre` optionally takes input file and output file arguments
- Improves how `cvtpcre` parses the PCRE test suites

### CODE REVIEW

**Comment by sfstewman** [comment]
I'm not 100% sure that this is ready to merge.  I need to write some test cases for a bunch of the grammar that's now supported or is now marked as unsupported.

**Comment by katef** [comment]
Am happy to merge this ahead of time, if the tests are coming later. Because this way around eases the work for merging, and then we don't end up blocked on this (excellent!) work, and also don't end up having to deal with merges on this side. I prefer merges to be handled wherever the conflicts are simpler.

**Comment by katef** [comment]
I also think we have more theft tests coming, and we'll pick up more of the pcre equivalence there too, eventually.

File: include/re/re.h
Comment (katef) [inline include/re/re.h:89]: amazing that a regexp dialect has a notion of comments, let alone ill-formed ones

File: src/libre/dialect/pcre/lexer.lx
Comment (katef) [inline src/libre/dialect/pcre/lexer.lx:51]: Excellent, thank you

File: src/libre/dialect/pcre/lexer.lx
Comment (katef) [inline src/libre/dialect/pcre/lexer.lx:36]: Ah the unsupported token makes sense. I hadn't imagined that!

File: src/libre/dialect/pcre/lexer.lx
Comment (katef) [inline src/libre/dialect/pcre/lexer.lx:110]: This works for me.

File: src/libre/dialect/pcre/lexer.lx
Comment (katef) [inline src/libre/dialect/pcre/lexer.lx:190]: This dialect needs a prop 65 warning

File: src/libre/parser.act
Comment (katef) [inline src/libre/parser.act:339]: If this is specific to pcre, we might want to split `CONTROL` into two tokens, and provide other behaviour for other dialects which use it. Thanks for the comment here, it's clear enough that we can see to do that in the future if neccessary.

Comment (katef) [reply src/libre/parser.act:339]: Do you think they designed this syntax, or was it machine evolved?

**Review by katef — APPROVED** [review]

**Comment by katef** [comment]
Brilliant work. Thank you again.

File: include/re/re.h
Comment (sfstewman) [reply include/re/re.h:89]: I actually like that it does.  People can write long, complex regexps, and comments along with the PCRE/Perl extended syntax makes them a lot more readable and maintainable.

**Pull request merged**
